File: src/main/java/io/github/seggan/jyxal/compiler/wrappers/InsnSequence.java
package io.github.seggan.jyxal.compiler.wrappers;

import org.objectweb.asm.tree.AbstractInsnNode;

import java.util.ArrayList;
import java.util.List;

record InsnSequence(int... insns) {

    public List<AbstractInsnNode> matches(AbstractInsnNode insn) {
        List<AbstractInsnNode> matches = new ArrayList<>();
        for (int i : insns) {
            if (insn.getOpcode() != i) {
                return new ArrayList<>();
            }
            matches.add(insn);
            insn = insn.getNext();
        }

        return matches;
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/compiler/wrappers/InsnSequence.kt
package io.github.seggan.jyxal.compiler.wrappers

import org.objectweb.asm.tree.AbstractInsnNode

internal data class InsnSequence(val insns: List<Int>) {

    constructor(vararg insns: Int) : this(insns.toList())

    fun matches(insnNode: AbstractInsnNode): List<AbstractInsnNode> {
        var insn = insnNode
        val matches = mutableListOf<AbstractInsnNode>()
        for (i in insns) {
            if (insn.opcode != i) {
                return ArrayList()
            }
            matches.add(insn)
            insn = insn.next
        }
        return matches
    }
}